vyper/semantics/types/module.py

```diff
@@ -0,0 +1,315 @@
+from functools import cached_property
+from typing import Optional
+
+from vyper import ast as vy_ast
+from vyper.abi_types import ABI_Address, ABIType
+from vyper.ast.validation import validate_call_args
+from vyper.exceptions import InterfaceViolation, NamespaceCollision, StructureException
+from vyper.semantics.analysis.base import VarInfo
+from vyper.semantics.analysis.utils import validate_expected_type, validate_unique_method_ids
+from vyper.semantics.namespace import get_namespace
+from vyper.semantics.types.base import TYPE_T, VyperType
+from vyper.semantics.types.function import ContractFunctionT
+from vyper.semantics.types.primitives import AddressT
+from vyper.semantics.types.user import EventT, StructT, _UserType
+
+
+class InterfaceT(_UserType):
+    _type_members = {"address": AddressT()}
+    _is_prim_word = True
+    _as_array = True
+    _as_hashmap_key = True
+    _supports_external_calls = True
+    _attribute_in_annotation = True
+
+    def __init__(self, _id: str, functions: dict, events: dict, structs: dict) -> None:
+        validate_unique_method_ids(list(functions.values()))
+
+        members = functions | events | structs
+
+        # sanity check: by construction, there should be no duplicates.
+        assert len(members) == len(functions) + len(events) + len(structs)
+
+        super().__init__(functions)
+
+        self._helper = VyperType(events | structs)
+        self._id = _id
+        self.functions = functions
+        self.events = events
+        self.structs = structs
+
+    def get_type_member(self, attr, node):
+        # get an event or struct from this interface
+        return TYPE_T(self._helper.get_member(attr, node))
+
+    @property
+    def getter_signature(self):
+        return (), AddressT()
+
+    @property
+    def abi_type(self) -> ABIType:
+        return ABI_Address()
+
+    def __repr__(self):
+        return f"interface {self._id}"
+
+    # when using the type itself (not an instance) in the call position
+    def _ctor_call_return(self, node: vy_ast.Call) -> "InterfaceT":
+        self._ctor_arg_types(node)
+        return self
+
+    def _ctor_arg_types(self, node):
+        validate_call_args(node, 1)
+        validate_expected_type(node.args[0], AddressT())
+        return [AddressT()]
+
+    def _ctor_kwarg_types(self, node):
+        return {}
+
+    # TODO x.validate_implements(other)
```

Comment:
What is this supposed to mean?

Reply:
i think at some point i wanted to reverse the order of the arguments. so the API looks like `other.implements(x)` as opposed to `x.implements(other)`